feat(init): Add MCP support to `rover init`

gocamille opened the pull request:
## Summary

This PR adds the `--mcp` flag to `rover init`, enabling three distinct flows to add the Apollo MCP server while maintaining complete backward compatibility and rollback safety. The implementation leverages Rover's existing TypeState architecture with dedicated MCP state transitions that are architecturally isolated but follow established patterns.

**Core Value:** Developers can now add AI tooling capabilities to their GraphQL projects with `rover init --mcp`, bridging the gap between GraphQL federation and AI development workflows.

**Review Note: This PR needs to be merged after the `add-mcp` template is merged here: https://github.com/apollographql/rover-init-starters/pull/77**

**Review Note: [The reference to `rover-init-starters`](https://github.com/apollographql/rover/blob/af994124d6f53ed75b5f5595297effa50d05d6c1/src/command/init/transitions.rs#L247) must be reverted before merging.**


## Implementation Strategy: TypeState Safety with Isolated Integration

Rather than disrupting the `rover init` TypeState machine, we've implemented MCP flows as **distinct state transitions** that branch early from the main flow. This approach provides isolation as all MCP logic is contained in a separate module, and can be completely reverted without touching existing flows. 

### The Two New MCP Flows

1. New Project + MCP Flow (`MCPSetupType::NewProject`)
- Creates new graph and project with MCP capabilities from scratch
- Uses template composition (base template + `add-mcp` template from rover-init-starters repository)
- Generates Claude Desktop configuration
- Follows full TypeState progression: `ProjectTypeSelected → MCPCreationConfirmed → ProjectCreated`

2. Existing Graph + MCP Flow (`MCPSetupType::ExistingGraph`)
- Adds MCP capabilities to existing project that references existing graph
- Leverages fuzzy search for graph selection (integrated from PR #2726)
- Bypasses graph creation, focuses on tooling setup
- Direct flow: `Authentication → Graph Selection → MCP Setup`

## Review Focus Areas

### Critical Path Review
1. `src/command/init/mod.rs`:89-95 - Early MCP branching logic
2. `src/command/init/transitions.rs`:566 - Guard clause protecting existing flows
3. Template composition logic - Ensure no conflicts with existing templates

### Architecture Review
1. State isolation - Verify MCP states don't interfere with existing TypeState machine
2. Error handling - Ensure MCP failures don't break existing error flows
3. Template merging - Validate `.gitignore` merging and variable substitution

####  Potential Rollback Strategy

**Immediate isolation points** (ordered by simplicity):
1. **Feature flag**: Add `mcp` feature flag - default builds exclude MCP entirely
2. **Early branch removal**: Delete 5 lines in `mod.rs:89-95` (the `if self.project_template.mcp` check)
3. **Module deletion**: Remove `src/command/init/mcp/` directory
4. **Full revert**: `git revert` this commit

**Feature flag recommendation**: Adding `mcp` to Cargo features allows production builds to completely exclude MCP code while enabling gradual rollout.

Note the early branching means rollback never touches existing TypeState machine - failure blast radius contained to new code only.

##  Implementation Deep Dive

#### 🔴 Major Updates - Core State Machine Changes (Focus Review Here)

**`src/command/init/mod.rs`**
- L89-95: Early MCP branch logic (bypasses directory validation)
- L97-180: New `handle_mcp_flow()` method with dedicated state transitions
- SAFETY: All existing code paths unchanged, new logic only triggers with `--mcp` flag

**`src/command/init/states.rs`**
- L90-140: New MCP-specific state types (`MCPSetupType`, `MCPAugmentationSelected`, etc.)
- SAFETY: Additive-only changes, no modifications to existing states

**`src/command/init/transitions.rs`**
- L1089-1200: New MCP state transition implementations
- L566: Guard clause preventing MCP flags from reaching normal template flow
- SAFETY: Early exit pattern protects existing transitions

####  🟡 Medium Risk Updates - Template System Enhancements

**`src/command/init/template_fetcher.rs`**
- L180-250: New `fetch_mcp_template()` method for template composition
- L290-340: `.gitignore` merging logic
- SAFETY: New functionality only, existing template fetching unchanged

**`src/command/init/template_operations.rs`**
- Enhanced hydration for MCP template variables
- SAFETY: Backward-compatible template variable handling

#### 🟢 Low Risk - New Functionality

**`src/command/init/mcp/`** (Entirely new module)
- `mcp_operations.rs`: version checking, Claude Desktop config generation
- Self-contained logic with no external dependencies

### New Test Coverage
- `tests/mcp_*_tests.rs`: 6 new test files with 40+ test cases
- Integration tests, behaviour tests, flow tests, transition tests
- Template consistency and hydration validation

 ### DEPENDENCY CHANGES - From Merged PRs

 **`src/command/output.rs`**
- Source: Merged PR #2722 (docs command refactoring)
- Impact: No interaction with MCP functionality
- Safety: Independent feature, tested separately

**src/command/docs/shortlinks.rs**
- Source: Merged PR #2722
- Impact: Documentation improvements only

**Fuzzy Select Integration**
- Source: Merged PR #2726
- Usage: Powers graph selection in MCP existing-graph flow
- Safety: Well-tested feature reused in new context

### INFRASTRUCTURE ADDITIONS

**`crates/rover-client/src/operations/init/list_graphs/`**
- New GraphQL operation for graph listing 
- Standard rover-client pattern: query + runner + types
- Safety: Read-only operation, no mutations

gocamille committed:
feat(init): MCP provisioning flow

feat(init): Refinements

feat(init): MCP refinements continued

feat(init): MCP refinements continued

feat(init): UX refinements for MCP flow

gocamille committed:
feat(init): Debug mcp server configuration

Commit:
feat(init): removing dead code and started moving some code around

linting

fixing tests

reformatting

gocamille committed:
feat(init): Refactor UX for mcp flow

feat(init): Minor UI updates to rover init mcp flow

gocamille committed:
Add bullet points to selection confirmations

gocamille committed:
feat(init): Additional UI updates to mcp flow

gocamille committed:
feat(init): Generate simple query for existing graphs

gocamille committed:
feat(init): update next steps and claude desktop configuration file generation

gocamille committed:
feat(init): remove local download of mcp server

gocamille committed:
feat(init): Update to UX prompt flow

gocamille committed:
feat(init): Re-introduce environment creation and update next steps prompt

feat(init): Update ux for existing projects running mcp provisioning

feat(init): resolve linting errors

tayrrible committed:
[feat]: Updating supergraph.yaml generation to handle changes to template files (#2721)

- Giving a hard default to subgraph naming for supergraph.yaml
generation to prevent panics
- Removing "invalid" option for selection for template selector
<img width="354" height="77" alt="image"
src="https://github.com/user-attachments/assets/5de441fb-f46b-4f71-b2f8-897dae0604a1"
/>
- Updating file names for template files


Dependent on [updates to
rover-init-starters](https://github.com/apollographql/rover-init-starters/pull/74)

gocamille committed:
feat(init): Refactor UX based on design critique

gocamille committed:
feat(init): add guide to include credentials in mcp config file

gocamille committed:
feat(init): Update the rover dev command for mcp flow

gocamille committed:
feat(init): Remove duplicate selections for rover init

feat(init): update wording for selecting graph ref

feat(init): fix lint errors

tayrrible committed:
[feat] Refactoring `docs` command to work with `ShortlinksInfo` (#2722)

Updated the `docs` command to allow us to expose doc links:
- not related to just Rover (allowing sub-routes)
- Separate "slug" from "key" to help make less fragile

Also required updating how output handled links.

esilverm committed:
feat: use fuzzy select for graph selection input (#2726)

<!--
First, 🌠 thank you 🌠 for taking the time to consider a contribution to
Apollo!

Here are some important details to follow:

* ⏰ Your time is important
To save your precious time, if the contribution you are making will
take more than an hour, please make sure it has been discussed in an
        issue first. This is especially true for feature requests!

* 💡 Features
Feature requests can be created and discussed within a GitHub Issue.
Be sure to search for existing feature requests (and related issues!)
prior to opening a new request. If an existing issue covers the need,
please upvote that issue by using the 👍 emote, rather than opening a
        new issue.

* 🕷 Bug fixes
These can be created and discussed in this repository. When fixing a
bug,
please _try_ to add a test which verifies the fix. If you cannot, you
should
still submit the PR but we may still ask you (and help you!) to create a
test.

* 📖 Contribution guidelines
Follow https://github.com/apollographql/rover/blob/HEAD/CONTRIBUTING.md
when submitting a pull request. Make sure existing tests still pass, and
add
        tests for all new behavior.

* ✏️ Explain your pull request
Describe the big picture of your changes here to communicate to what
        your pull request is meant to accomplish. Provide 🔗 links 🔗 to
        associated issues!

We hope you will find this to be a positive experience! Open source
contribution can be intimidating and we hope to alleviate that pain as
much
as possible. Without following these guidelines, you may be missing
context
that can help you succeed with your contribution, which is why we
encourage
discussion first. Ultimately, there is no guarantee that we will be able
to
merge your pull-request, but by following these guidelines we can try to
avoid disappointment.

-->
Addresses AIR-26. Graph selection workflow utilizes fuzzy selection
instead of normal selection workflow to enable narrowing of options

gocamille committed:
feat(init): Ensure arguments are read when running rover init --mcp

gocamille committed:
feat(init): Ensure .env.template file is converted to .env file with credentials

gocamille committed:
feat(init): Update file groupings listed in confirmation step

gocamille committed:
feat(init): Update copy and normalize docker name

gocamille committed:
feat(init): Update next step command for users adding mcp to existing project

gocamille committed:
feat(init): Remove tools directory generation from MCP projects

feat(init): Remove unused variable

gocamille committed:
feat(init): Add graph_id prompt for new projects

gocamille committed:
feat(init): Remove line break for prompt

gocamille committed:
feat(init): Update command to list mcp docs

gocamille committed:
feat(init): Refactor mcp flows to follow TypeState architecture

gocamille committed:
feat(init): Add tests for mcp flows

gocamille committed:
feat(init): Remove unused variables and functions

feat(init): Resolve linting errors

gocamille commented on src/command/init/transitions.rs:
```diff
@@ -238,26 +242,48 @@ impl UseCaseSelected {
         // Fetch the template to get the list of files
-        let repo_ref = "release/v2";
-        let template_fetcher = InitTemplateFetcher::new().call(repo_ref).await?;
+        let repo_ref = "camille/start-with-mcp-template";
```
**Review Note: [The reference to `rover-init-starters`](https://github.com/apollographql/rover/blob/af994124d6f53ed75b5f5595297effa50d05d6c1/src/command/init/transitions.rs#L247) must be reverted before merging.**

gocamille commented on src/command/init/transitions.rs:
```diff
@@ -501,3 +639,511 @@ impl Completed {
+        };
+
+        // TODO: Remove hardcoded repo_ref and implement proper template composition
+        let repo_ref = "camille/start-with-mcp-template";
```
**Review Note: [The reference to `rover-init-starters`](https://github.com/apollographql/rover/blob/af994124d6f53ed75b5f5595297effa50d05d6c1/src/command/init/transitions.rs#L247) must be reverted before merging.**

gocamille committed:
feat(init): Mark mock secrets in tests as safe

tayrrible commented on src/command/docs/shortlinks.rs:
```diff
@@ -1,32 +1,123 @@
-pub const URL_BASE: &str = "https://go.apollo.dev/r";
+pub const URL_BASE: &str = "https://go/apollo/dev";
```
@gocamille I have a fix for this on my pr

tayrrible commented on src/command/docs/shortlinks.rs:
```diff
@@ -46,20 +137,54 @@ mod tests {
     #[test]
     fn can_get_url_from_slug() {
-        let expected_link = "https://go.apollo.dev/r/start";
+        let expected_link = "https://go/apollo/dev/r/start";
```
lol idk if i fixed this one on my branch
```suggestion
        let expected_link = "https://go.apollo.dev/r/start";
```

tayrrible replied:
Okay, fixed this on my branch.

gocamille replied:
No worries -- I will investigate!

gocamille committed:
feat(init): Add agent rules during MCP provisioning

tayrrible commented on .gitleaks.toml:
```diff
@@ -6,4 +6,16 @@
+      "src/command/init/tests/mcp_next_steps_test.rs",
+      "src/command/init/tests/template_hydration_tests.rs",
+      "src/command/init/tests/mcp_prompting_tests.rs",
+      "src/command/init/mod.rs",
```
Just curious, why are we allowing it in mod.rs? Do we have a test that uses it?

gocamille replied:
We have copy that is included in the "Next Steps" section when someone runs `rover init --mcp` that refers to a user's key and provides an example, but I agree I do _not_ want to allowlist this, just the tests. I have reverted to having the `-gitleaks-allow` inline flag mark that specific line as allowable, and removed this file whitelisting 👍

tayrrible committed:
feat: Updating error messages, updating templates with generated creds, updating `docs` command (#2732)

- Updating error messages for users running command in non-empty
directories/paths
- Updating the templates to replace template values with values
generated during `init`
  - including APOLLO_KEY & APOLLO_GRAPH_REF
- Updating the `docs` command to work with non-rover specific links

tayrrible commented on src/command/init/transitions.rs:
```diff
@@ -238,26 +242,48 @@ impl UseCaseSelected {
+        // MCP flow is handled in separate state machine, should not reach here with --mcp flag
+        if options.mcp {
+            return Err(RoverError::new(anyhow!(
+                "Internal error: MCP flow should use dedicated state transitions"
```
@gocamille copilot calling this out.  is there a different error we should be using? this seems like an internal to us error and not something the user can take action against

gocamille replied:
Yes I agree -- this doesn't seem like something we need to take action against.

tayrrible commented on src/command/init/transitions.rs:
```diff
@@ -480,16 +509,179 @@ impl CreationConfirmed {
+        println!("{}", Style::Heading.paint("Next steps ↴"));
+        println!();
+
+        // Step 1: Always Claude Desktop configuration first
+        println!("1. Configure Claude Desktop by copying claude_desktop_config.json to:");
+        println!(
+            "   • macOS: {}",
+            Style::Path.paint("~/Library/Application Support/Claude/claude_desktop_config.json")
+        );
+        println!(
+            "   • Windows: {}",
+            Style::Path.paint("%APPDATA%\\Claude\\claude_desktop_config.json")
+        );
+        println!(
+            "   • Linux: {}",
+            Style::Path.paint("~/.config/Claude/claude_desktop_config.json")
+        );
+        println!();
+        println!("   Then restart Claude Desktop.");
+        println!();
```
I'm a little nervous about hardcoding this into the rover version on the chance that Claude desktop changes something on their side 🤔

gocamille replied:
I think our only other option is to refer to the docs: https://www.apollographql.com/docs/apollo-mcp-server/quickstart#step-4-connect-claude-desktop although that would take someone out of the flow, and we'd also have to look at adding it to `rover docs open` (unless we're okay with hardcoding this url for today's release). Thoughts @tayrrible ?

tayrrible replied:
I think we can leave as is for this release but should consider updating in a future release

tayrrible commented on src/command/init/transitions.rs:
```diff
@@ -480,16 +509,179 @@ impl CreationConfirmed {
+        );
+        println!(
+            "- Learn more → {}",
+            Style::Command.paint("rover docs open mcp-qs")
```
is the quickstart the right link for this?

gocamille replied:
This one takes people to the MCP quickstart: https://www.apollographql.com/docs/apollo-mcp-server/quickstart -- was it the wording that you were more concerned about, or the functionality (are you having trouble when running `rover docs open mcp-qs`)?

tayrrible committed:
fixing lint errors

tayrrible commented on src/command/init/mod.rs:
```diff
@@ -192,6 +310,1310 @@ impl Init {
+        let selection = FuzzySelect::new()
+            .with_prompt(
+                Style::Prompt
+                    .paint("? Select existing graph variant to work with (or type graph name):"),
```
enter the graph name? or graph ref? or graph id?

gocamille replied:
since it's fuzzysearch, the graph name will give them all graphs that comprise their search value -- but we can say graph id instead 👍

tayrrible replied:
i think that's fine as is then! ty!

tayrrible commented on src/command/init/mod.rs:
```diff
@@ -192,6 +310,1310 @@ impl Init {
+        if all_graph_options.is_empty() {
+            println!("{}", Style::WarningHeading.paint("▲ No graphs found"));
+            println!(
+                "You need to create a graph first. Visit https://studio.apollographql.com to create your first graph."
```
should we just tell them to visit studio to connect an existing graph or run init to create a new one?

gocamille replied:
I think telling them to run `rover init` first is an excellent idea and I'd swap that out for the suggestion to go to Studio

sanchezdale commented on src/command/init/options/project_template.rs:
```diff
@@ -1,64 +1,45 @@
-                )));
-            }
+    /// Get base template ID from MCP template ID
+    /// DEPRECATED: MCP flow now uses dedicated state transitions
```
nit: can we remove this before merge

gocamille committed:
chore: point to latest rover-init-starters template branch

gocamille committed:
chore: fix linter issues

gocamille committed:
chore: remove deprecated code

gocamille committed:
Remove file-level allowlist of mod.rs file

gocamille committed:
feat(init): Update internal error

gocamille committed:
feat(init): Update prompt for searching for graph by id

gocamille committed:
feat(init): Update copy for error on empty directory

gocamille committed:
feat(init): Update macos runners in CI/CD to match what our Router uses

tayrrible committed:
swapping messaging and formatting

gocamille committed:
feat(init): Add another commit to whitelist

gocamille committed:
feat(init): Add another commit to whitelist

gocamille committed:
feat(init): Add test files to whitelist
